src/librustc_middle/traits/mod.rs

```diff
@@ -102,15 +128,16 @@ impl<'tcx> ObligationCause<'tcx> {
         body_id: hir::HirId,
         code: ObligationCauseCode<'tcx>,
     ) -> ObligationCause<'tcx> {
-        ObligationCause { span, body_id, code }
+        ObligationCause { data: Lrc::new(ObligationCauseData { span, body_id, code }) }
     }
 
     pub fn misc(span: Span, body_id: hir::HirId) -> ObligationCause<'tcx> {
-        ObligationCause { span, body_id, code: MiscObligation }
+        ObligationCause::new(span, body_id, MiscObligation)
     }
 
-    pub fn dummy() -> ObligationCause<'tcx> {
-        ObligationCause { span: DUMMY_SP, body_id: hir::CRATE_HIR_ID, code: MiscObligation }
+    #[inline(always)]
+    pub fn dummy(tcx: TyCtxt<'tcx>) -> ObligationCause<'tcx> {
+        tcx.obligation_causes.dummy.clone()
```

Comment:
We should use a `OnceCell<Lrc<ObligationCause<'static>>>` for this instead of passing `TyCtxt` everywhere.